### PR TITLE
Use pyproject.toml to configure tools, rather than tox.ini + mypy.ini

### PR DESCRIPTION
**Pull request by jnsgruk**
For discussion: we could consider using `pyproject.toml` to configure the various linters, formatters and testing tools. We were split across `tox.ini` and `mypy.ini` already, and this approach has some other interesting characteristics:

- People who want to *not* use `tox` for whatever reason can invoke `flake8`, `black`, `mypy` separately and the configs should be picked up automatically
- All config is centralised
- Declutter the `tox.ini` and make it more single-purpose (i.e. define test environments and commands)
- Add a `vars` section to the `tox.ini` that contains commonly used paths, this should make it easier to reuse this file across out projects.

One drawback here: `flake8` doesn't yet natively support using `pyproject.toml`, so we have to use a third-party package: `pyproject-flake8`. This doesn't have any huge impact, other than requiring `flake8` to be invoked using a wrapped named `pflake8`. This does sort of negate my first point above, but if people follow the `CONTRIBUTING.md` docs, it shouldn't be a problem.

### CODE REVIEW

**Review by balbirthomas — APPROVED** [review]

File: tox.ini [hunk 1]
```diff
@@ -68,12 +43,13 @@ deps =
...
     flake8-copyright
     flake8-builtins
+    pyproject-flake8
     pep8-naming
     isort
 commands =
-    flake8 {toxinidir}/src {toxinidir}/tests {toxinidir}/lib/charms/alertmanager_k8s
-    isort --check-only --diff {toxinidir}/src/. {toxinidir}/tests/. {toxinidir}/lib/charms/alertmanager_k8s/.
-    black --check --diff --line-length=99 --target-version=py38 {toxinidir}/src {toxinidir}/tests {toxinidir}/lib/charms/alertmanager_k8s
+    pflake8 {[vars]all_path}
```
Comment (sed-i) [inline tox.ini:50]: Add comment about (temporary) use of pflake?

File: tox.ini [hunk 2]
```diff
@@ -1,15 +1,20 @@
...
 
+[vars]
+src_path = {toxinidir}/src/.
+tst_path = {toxinidir}/tests/.
+lib_path = {toxinidir}/lib/charms/alertmanager_k8s/.
+all_path = {[vars]src_path} {[vars]tst_path} {[vars]lib_path}
+
 [testenv]
-basepython = python3.8
+basepython = python3
```
[the remark this answers is not in the document]
Comment (dstathis) [reply tox.ini:15]: I think it's better to leave this as 3. We will all be on 3.9 in the not too distant future and I would be surprised if we remember to change this.

File: tox.ini [hunk 3]
```diff
@@ -87,7 +63,7 @@ deps =
     types-setuptools
     types-toml
 commands =
-    mypy --config-file={toxinidir}/mypy.ini {toxinidir}/src/charm.py {toxinidir}/tests {posargs}
+    mypy {[vars]src_path}/charm.py {posargs}
```
[the remark this answers is not in the document]
Comment (dstathis) [reply tox.ini:66]: Yes this should definitely be checking the tests as well as the library

File: tox.ini [hunk 4]
```diff
@@ -1,15 +1,20 @@
...
-
 [tox]
 skipsdist=True
 skip_missing_interpreters = True
 envlist = lint, static, unit
 
+[vars]
+src_path = {toxinidir}/src/.
+tst_path = {toxinidir}/tests/.
+lib_path = {toxinidir}/lib/charms/alertmanager_k8s/.
```
[the remark this answers is not in the document]
Comment (dstathis) [reply tox.ini:11]: I would be shocked if you need that "." though I suppose it shouldn't really hurt anything.

**Review by dstathis — DISMISSED** [review]
This is a great change. Only requesting changes because mypy needs to run against the lib and tests.

File: tox.ini (hunk 2 above)
[the remark this answers is not in the document]
Comment (jnsgruk) [reply tox.ini:15]: Yeh, my machine (`hirsute`) is 3.9 and it just fails out with a missing interpreter if you specify 3.8 :)

File: tox.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (jnsgruk) [reply tox.ini:11]: You need it for `isort`, and from my testing today it doesn't negatively impact the other tools.

File: tox.ini (hunk 1 above)
Comment (jnsgruk) [reply tox.ini:50]: Will do

File: tox.ini (hunk 2 above)
[the remark this answers is not in the document]
Comment (rbarry82) [reply tox.ini:15]: That said, OF supports 3.6+, and running a suite across supported versions could be nice to catch new syntax which fails on older versions (some f-strings, walrus, etc).

File: tox.ini (hunk 4 above)
[the remark this answers is not in the document]
Comment (jnsgruk) [reply tox.ini:11]: hmm, seems you can get away without it for `isort`. Will fix in next commit

**Comment by jnsgruk** [comment]
If you're happy @sed-i, merge when you're ready and we can get this out across the other charms. If you take your other two, I'll roll some PRs tomorrow for other charms :)

Note: latest commit added some annotations to your tests where a specific mypy check was failing.

File: tox.ini (hunk 2 above)
[the remark this answers is not in the document]
Comment (sed-i) [reply tox.ini:15]: The reason I have 3.8 there is that k8s charms must be runnable with 3.8.
If the template is ported to ops, then I would set it to 3.6.
In GH workflows we install the python version(s) we need.
IIUC, which interpreter(s) happens to be available locally should be irrelevant, i.e. just specifying "python3" seems to be unmindful.

[the remark this answers is not in the document]
Comment (jnsgruk) [reply tox.ini:15]: I'm not going to fight this either way, happy for this to be set back to 3.8, but perhaps include a note in CONTRIBUTING.md, as the suite will fail to run on machines that don't have python3.8 in the path if no other changes are made :)

[the remark this answers is not in the document]
Comment (jnsgruk) [reply tox.ini:15]: Hmm, actually, I'd really prefer to keep this as it is.

The point is we can easily catch failures for existing versions as part of a test suite on GitHub, but specifying a version here is likely to trip up developers locally, and it's not super trivial to fix.

I think leave this, and make sure the test environment on your actions is the right version, given that it must pass before merge anyway, we should be covered.

**Review by sed-i — APPROVED** [review]

**Pull request merged**